File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, nodeA, nodeB):
            if not nodeA and not nodeB:
                return True
            elif nodeA and nodeB and nodeA.val == nodeB.val:
                return self.isSameTree(nodeA.left, nodeB.left) and self.isSameTree(nodeA.right, nodeB.right)
            else:
                return False

    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:
            return True
        
        if not root:
            return False
        
        if self.isSameTree(root, subRoot):
            return True
        
        left = self.isSubtree(root.left, subRoot)
        right = self.isSubtree(root.right, subRoot)
        
        return left or right

File: Data Structures & Algorithms/balanced-binary-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {


    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }

        if (Math.abs(this.dfs(root.left) - this.dfs(root.right)) > 1) {
            return false;
        }

        return this.isBalanced(root.left) && this.isBalanced(root.right);
        
    }

    public int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int left = this.dfs(node.left);
        int right = this.dfs(node.right);

        return 1 + Math.max(left, right);
    }


}
